proposals/per_policy_webhook_configurations.md

```diff
@@ -0,0 +1,235 @@
+# Meta
+[meta]: #meta
+- Name: Policy-based Webhook Configurations
+- Start Date: 2023-12-12
+- Update data (optional):
+- Author(s): @realshuting
+
+# Table of Contents
+[table-of-contents]: #table-of-contents
+- [Meta](#meta)
+- [Table of Contents](#table-of-contents)
+- [Overview](#overview)
+- [Definitions](#definitions)
+- [Motivation](#motivation)
+- [Proposal](#proposal)
+- [Implementation](#implementation)
+- [Migration (OPTIONAL)](#migration-optional)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Prior Art](#prior-art)
+- [Unresolved Questions](#unresolved-questions)
+- [CRD Changes (OPTIONAL)](#crd-changes-optional)
+
+# Overview
+[overview]: #overview
+
+Configure Kyverno resource webhook configurations based on policy settings.
+
+# Definitions
+[definitions]: #definitions
+
+Kyverno has two webhook configurations registered for the selected resources by policies, a MutatingWebhookConfiguration and a ValidatingWebhookConfiguration respectively. For brevity, this proposal simply uses "webhooks" to refer to the two mentioned webhook configurations.
+
+# Motivation
+[motivation]: #motivation
+
+From [#9111](https://github.com/kyverno/kyverno/issues/9111) and [#8063](https://github.com/kyverno/kyverno/issues/8063), the request is to register webhooks for the policy selected resources which:
+
+1. allows users to customize webhooks flexibly 
+2. eliminates unnecessary admission requests being forwarded to Kyverno
+
+Benefits of policy based webhooks:
+1. offers the flexibility for webhook configurations
+2. prevents unexpected webhook failures
+3. reduces the admission review processing duration
+
+# Proposal
+
+There are two dimensions for this feature:
+
+1. offer users the option to configure webooks' attributes including `matchConditions`, `objectSelector` and `namespaceSelector` based on the policy settings
+2. register webhooks based on the policy type, specifically to configure the `scope` attribute of webhooks for namespaced policies
+
+The following snippets list out how policy configurations are transformed to the webhook configurations.
+
+**a. selectors that filters resource objects**
+
+The existing attribute `spec.rules.match.(any/all).resources.selector` can be transformed to the webhooks `objectSelector`.
+```yaml
+spec:
+  rules:
+    - match:
+        all:
+        - resources:
+            selector:
+              matchLabels:
+                foo: bar
+```
+The above rule will be transformed to the following webhook configuration.
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+- name: my-webhook.example.com
+  objectSelector:
+    matchLabels:
+      foo: bar
```

Comment:
Is it possible to use set-based filtering here? (ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#list-and-watch-filtering)

Otherwise I think this will be messy IMO, constructing an "OR" for these two policy cases:
```
spec:
  rules:
    - match:
        all:
        - resources:
            selector:
              matchLabels:
                foo: bar
```
```
spec:
  rules:
    - match:
        all:
        - resources:
            selector:
              matchLabels:
                foo: foo
```

Would end up in two unique webhook configurations in the ValidatingWebhookConfiguration -  or alternatively only `matchConditions` should be used

Reply:
Another thing is what happens if we have a policy that matches all resources with a specific label and another one that matches a configmap for example? We will end up have a webhook configuration that matches configmaps with this label.

Example:
1.  The first policy matches all resources with a specific label:
```
spec:
  rules:
    - match:
        all:
        - resources:
            selector:
              matchLabels:
                foo: bar
```
2. The second policy matches configmaps:
```
rules:
  - match:
      any:
      - resources:
          kinds:
          - ConfigMap
```

The generated configuration will be as follows:
```
  objectSelector:
    matchLabels:
      foo: bar
  rules:
  - apiGroups:
    - ""
    apiVersions:
    - v1
    operations:
    - CREATE
    - UPDATE
    - DELETE
    - CONNECT
    resources:
    - configmaps
   ```
Therefore, configmaps with this label will only be forwarded to the webhook server.

Reply:
It seems to me like our best course of action is to have a hierarchy for which changes are made to the webhook config. So if e.g. some policy has `any resource` and another has `resource with label` then our logic should decide that `any resource` > `resource with label`.
Then only the rule without objectselector should be added.

It sounds easy but I expect this to be reasonably complex.

Reply:
I wonder if we could start with the `matchConditions` and make the webhook rule and kyverno rule 1-1 map. This way it requires users to configure webhook filters with caution.